File: src/solutions/2024/01/csharp/Main.cs

using System;
using Core;

namespace Year2024;
public static class Day01
{
   static SolutionOptions OPTIONS = new()
   {
      HasAlternate = false,
      HasIO = false,
   };

   static string Solve(string input, bool p2)
   {
      var l = new List<int>();
      var r = new List<int>();
      var hashmap = new Dictionary<int, int>();
      foreach (var line in input.Split('\n'))
      {
         var pair = line.Split("   ");
         var k = int.Parse(pair[0]);
         var v = int.Parse(pair[1]);
         l.Add(k);
         r.Add(v);
         hashmap.TryAdd(k, 0);
         hashmap.TryAdd(v, 0);
         hashmap[v]++;
      }
      l.Sort();
      r.Sort();

      int sum = 0;
      if (p2)
      {
         foreach (var e in l)
         {
            sum += e * hashmap[e];
         }
      }
      else
      {
         for (int i = 0; i < l.Count; i++)
         {
            sum += Math.Abs(l[i] - r[i]);
         }
      }

      return sum.ToString();
   }

   static string Part1(string input, bool isTest = false)
   {
      return Solve(input, false);
   }

   static string Part2(string input, bool isTest = false)
   {
      return Solve(input, true);
   }

   static void Main(string[] args)
   {
      Runner.Run(args, Part1, Part2, OPTIONS);
   }
}

File: src/solutions/2024/01/ts/main.ts
import { argv } from 'node:process';
import type { SolutionOptions } from 'src/options.ts';
import { run } from 'src/run.ts';

const options: SolutionOptions = {
   hasAlternate: false,
   hasIo: false,
};

function solve(input: string, p2: boolean): string {
   const l: number[] = [];
   const r: number[] = [];
   const hashmap: { [key: number]: number } = {};
   input.split('\n').forEach((line) => {
      const pair = line.split('   ');
      const k = +pair[0];
      const v = +pair[1];
      l.push(k);
      r.push(v);
      hashmap[k] ||= 0;
      hashmap[v] ||= 0;
      hashmap[v]++;
   });
   l.sort((a, b) => a - b);
   r.sort((a, b) => a - b);

   let sum = 0;
   if (p2) {
      for (const e of l) {
         sum += e * hashmap[e];
      }
   } else {
      for (let i = 0; i < l.length; i++) {
         sum += Math.abs(l[i] - r[i]);
      }
   }

   return sum.toString();
}

function part1(input: string, _isTest: boolean): string {
   return solve(input, false);
}

function part2(input: string, _isTest: boolean): string {
   return solve(input, true);
}

if (import.meta.main) {
   run(argv, part1, part2, options);
}
